src/main/java/com/hierynomus/smbj/share/NamedPipe.java

```diff
@@ -141,7 +196,15 @@ public int transact(byte[] inBuffer, byte[] outBuffer) {
      * @return the number of bytes written to <code>outBuffer</code>
      */
     public int transact(byte[] inBuffer, int inOffset, int inLength, byte[] outBuffer, int outOffset, int outLength) {
-        return ioctl(FSCTL_PIPE_TRANSCEIVE, true, inBuffer, inOffset, inLength, outBuffer, outOffset, outLength);
+        if (outBuffer.length - outOffset < outLength) {
+            throw new IllegalArgumentException();
+        }
+        final byte[] transactOutBuffer = transact(inBuffer);
+        if (transactOutBuffer.length > outLength) {
+            throw new BufferOverflowException();
```

Comment:
\1 We can guard against garbage with the synchronize keyword on the transactFully method.  We need this anyway since it's possible for a multi threaded application to use the named pipe and unintentionally run into an in-progress transact.  They either wrap it with synchronization, or we do it here.
\2 Agree, the input stream does that in the current implementation.
\3 Why do you want to peek the stream?

Your suggestions would hide direct access to the API.  It also creates a state object (NamedPipeTransaction) when the NamedPipe object could be that state since it can not support parallel transactions.